common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java

```diff
@@ -119,6 +83,52 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
         return exceptionFromTokenResult(result.getTokenResult(), commandParameters);
     }
 
+    public static BaseException exceptionFromAuthorizationResult(final AuthorizationResult authorizationResult, final CommandParameters commandParameters) {
+        final String methodTag = TAG + ":exceptionFromAuthorizationResult";
+        final AuthorizationErrorResponse authorizationErrorResponse = authorizationResult.getAuthorizationErrorResponse();
+        //THERE ARE CURRENTLY NO USAGES of INVALID_REQUEST
+        switch (authorizationResult.getAuthorizationStatus()) {
+            case FAIL:
+                // Check if the error is to register device and throw DEVICE_REGISTRATION_NEEDED exception
+                if (authorizationErrorResponse instanceof MicrosoftAuthorizationErrorResponse) {
+                    MicrosoftAuthorizationErrorResponse microsoftAuthorizationErrorResponse =
+                            (MicrosoftAuthorizationErrorResponse) authorizationErrorResponse;
+
+                    if (microsoftAuthorizationErrorResponse.getError().equals(
+                            MicrosoftAuthorizationErrorResponse.DEVICE_REGISTRATION_NEEDED)) {
+
+                        return new DeviceRegistrationRequiredException(
+                                microsoftAuthorizationErrorResponse.getError(),
+                                microsoftAuthorizationErrorResponse.getErrorDescription(),
+                                microsoftAuthorizationErrorResponse.getUpnToWpj()
+                        );
+                    }
+                }
+
+                return new ServiceException(
+                        authorizationErrorResponse.getError(),
+                        authorizationErrorResponse.getErrorDescription(),
+                        ServiceException.DEFAULT_STATUS_CODE,
+                        null
+                );
+
+            case SDK_CANCEL:
+                return new ClientException(
+                        authorizationErrorResponse.getError(),
+                        authorizationErrorResponse.getErrorDescription()
+                );
+
+            case USER_CANCEL:
+                return new UserCancelException();
+            default:
+                Logger.warn(
+                        methodTag,
+                        "No AuthorizationResult status set"
+                );
+                return new ClientException(ClientException.UNKNOWN_ERROR, "Authorization failed with unknown error");
```

Comment:
We should include the error & errorDescription information from the response in the exception here, or is that passed in some other form to the caller?